tabbycat/draw/manager.py

```diff
@@ -197,9 +198,18 @@ def get_teams(self):
         metrics = self.round.tournament.pref('team_standings_precedence')
         pullup_metric = PowerPairedDrawGenerator.PULLUP_RESTRICTION_METRICS[self.round.tournament.pref('draw_pullup_restriction')]
 
-        generator = TeamStandingsGenerator(metrics, ('rank', 'subrank'), tiebreak="random",
-            extra_metrics=(pullup_metric,) if pullup_metric and pullup_metric not in metrics else ())
-        standings = generator.generate(teams, round=self.round.prev)
+        if self.round.prev is None:
+            standings = Standings(teams)
+            standings.record_added_metric('points', _("points"), _("Pts"), None, False)
+            standings.record_added_metric('seed', _("seed"), _("Seed"), None, False)
+            for team in teams:
+                standings.add_metric(team, 'points', 0)
+                standings.add_metric(team, 'seed', team.seed)
+            standings.sort(('points', 'seed'), random.shuffle)
+        else:
+            generator = TeamStandingsGenerator(metrics, ('rank', 'subrank'), tiebreak="random",
+                extra_metrics=(pullup_metric,) if pullup_metric and pullup_metric not in metrics else ())
+            standings = generator.generate(teams, round=self.round.prev)
```

Comment:
_[Scratch this comment and replace with the next one below, but I'm leaving this here so that my progression of thinking / brainstorm is open.]_

Would there be less special-case logic if we just appended a `'seed'` team metric to the metrics lists passed into `TeamStandingsGenerator`? So that it kind of becomes another tiebreak? (Though not in the same way—I realise we'd still have to fall back to `random.shuffle` because seeds aren't unique.) This implementation of diving back into low-level functions to generate a "special standings" seems kinda clunky and non-extensible.

I guess in this case we'd need to configure it using a new `Round` property that controls whether to append `'seed'`, since we don't want the seed to be in the actual rankings. But I was thinking we don't want to be tied specifically to the first round, if we can use a structure that is both simpler in implementation and more flexible to use.

(Admittedly I sometimes wondered whether the "standings precedence for the draw" and "standings precedence for the published tab" should be separated... but that's another topic.)

_[Again, scratch this comment and replace with the one below.]_

Reply:
Actually, a possibly better idea: Rather than implementing logic that treats `self.round_prev is None` as a special case of power-paired draws (since I'd argue it's not really, and implementing it as one is a bit of a kludge), write a new `DrawGenerator` subclass that just does what we want directly.

This new `SeededDrawGenerator` should just strip out everything that's not needed to keep the logic clear—skip all of the complicated odd bracket resolution logic, skip the `points` field, maybe even skip the pairing function options. It should just implement the seed pairing (as desired by APDA) directly. (Or possibly extract common logic to a common parent class with `PowerPairedDrawGenerator`.)

To select it, add a new `Round.draw_type` option, so the first round would have draw type "seeded" rather than "random".

Usability downside: APDA tournaments would need to edit the first round to choose this option explicitly (or we'd need to make some UI to make it easy, _e.g._ by allowing the user to select it on the availability page.

Reply:
I would make a new draw manager for seeds, but I wouldn't want to make a new draw generator, as the complexity it would add between the new mixins and inherited classes would overweigh the convenience of the generator having implemented extra steps (that would be skipped anyhow) for a more general case. It's already somewhat hard to follow between generators and managers, so subdividing even more wouldn't help.

Reply:
_[comment written before reviewing new code]_

Hmm, yeah, when I decided to rely on mixins and subclasses to mix and match functionality, I was mostly inspired by the patterns they use for class-based views in Django. I'd be open to the idea that this architecture is poorly suited for this application.

Brainstorm idea: Would it make sense to conceptualise draw generation as a series of steps (like, transforms), that are built into a sequence? So then each step would be in its own class, and some sort of composite class would know how to take in a list of steps and apply each one in turn. Examples of steps would be each existing option for each existing step:

- pull up teams to resolve odd brackets
- create intermediate brackets to resolve odd brackets
- rank by seed within brackets
- rank by points within brackets
- pair teams by folding within brackets

Major challenge would be defining a (common) interface for steps, because the input and output of each step don't "naturally" look the same. Some will output lists of teams in brackets, some will guarantee that brackets are aff/neg balanced, others will output an actual draw.

This might be too major a refactor to be worth it in this PR... but I wanted to put the thought up for discussion somewhere. Or another architecture, open to suggestions.